### ISSUE TITLE
`scipp.mod` ignores the `divisor`'s units

### ISSUE THREAD

**Issue by g5t** [scipp/scipp#3402]
The modulus operation is the [remainder after floor division](https://scipp.github.io/generated/functions/scipp.mod.html#scipp.mod) and should therefore require compatible units between its `dividend` and `divisor`.

For values with compatible units
```python
import scipp
dividend = scipp.scalar(168666898, unit='ns')
divisor = 1 / (14.0 * scipp.Unit('Hz'))

print(f'{dividend % divisor = :c}')
```
I would expect the modulus to 'just work', but instead the result is `dividend % divisor = 0.0 ns`

Forcing the values to have the _same_ units produces the expected result

`dividend % divisor.to(unit="ms") = 40.857142865657806 ns`

**Comment by jl-wynen** [comment]
The current behaviour is actually consistent with the definition:
```python
import scipp as sc
from scipp.testing import assert_identical

dividend = sc.scalar(168666898, unit='ns')
divisor = 1 / (14.0 * sc.Unit('Hz'))
f = dividend // divisor
r = dividend % divisor
print(f.value, f.unit, '|', r.value, r.unit)
assert_identical(f * divisor + r, dividend.to(dtype='float64'))

divisor = divisor.to(unit='ns')
f = dividend // divisor
r = dividend % divisor
print(f.value, f.unit, '|', r.value, r.unit)
assert_identical(f * divisor + r, dividend.to(dtype='float64'))
```
prints
```
2361336572.0 1e-09 | 0.0 ns
2.0 dimensionless | 25809755.142857164 ns
```
and the assertions pass.

The reason that the remainder differs in the two cases is that the quotient is also different. The remainder is defined wrt to the integer part of the dividend and divisor. Doing a unit conversion changes that integer part. This is an unfortunate interplay of units and scales but I don't think there is anything we can do differently. You simply have to be mindful of the scales involved and that the remainder is only defined in the context of a specific quotient.

**Comment by SimonHeybrock** [comment]
How about requiring identical units for `mod`, i.e., raise unless both are `ns`? I think everything else is too confusing.

**Comment by jl-wynen** [comment]
This would prevent dividing by dimensionless or counts which are both quite natural, IMHO

**Comment by g5t** [comment]
Maybe there is another conceptual interpretation of what the modulus operator does, but my handle on it aligns with [Modular arithmetic](https://en.wikipedia.org/wiki/Modular_arithmetic). Specifically, the idea of [congruence](https://mathworld.wolfram.com/Congruence.html) 

> If two numbers b and c have the property that their difference b-c is integrally divisible by a number m (i.e., (b-c)/m is an integer), then b and c are said to be "congruent modulo m."
> -- Wolfram MathWorld

With emphasis on _'integrally divisible'_.

If the numbers `b`, `c` and `m` have units, the only way for `(b-c)/m` to be an integer is if all numbers have the same units.

Under this interpretation, it _could_ make sense for `scipp.mod` to convert the units of the divisor or dividend such that they are the same; but like the difference between [`numpy.mod`](https://numpy.org/doc/stable/reference/generated/numpy.mod.html) and [`numpy.fmod`](https://numpy.org/doc/stable/reference/generated/numpy.fmod.html) when it comes to the _sign_ of the result, there would be an implementation choice for which _unit_ the result has.
Instead @simonheybrok's suggestion to require identical units would force the user to decide what the output units will be.

**Comment by jl-wynen** [comment]
> If the numbers b, c and m have units, the only way for (b-c)/m to be an integer is if all numbers have the same units.

Not really. You can have integers with units. E.g. `(10counts - 6counts) / 2`

**Comment by SimonHeybrock** [comment]
> This would prevent dividing by dimensionless or counts which are both quite natural, IMHO

Can you explain?

**Comment by jl-wynen** [comment]
> Can you explain?

If you have an object that is 6m long and want to divide it into 5 equally long pieces and want to length of the leftover, you do `6m % 5`.

**Comment by nvaytet** [comment]
I think I always understood it as "I want to make pieces of 5m, how much would I be left with?"

**Comment by jl-wynen** [comment]
But is my use case invalid? It would be with your suggestion.

**Comment by SimonHeybrock** [comment]
I think of "mod" as "loop around in a ring". From that point of view I feel "6m % 5" is weird. In fact I have still no idea what the unit of that should be.

**Comment by nvaytet** [comment]
> If you have an object that is 6m long and want to divide it into 5 equally long pieces and want to length of the leftover, you do 6m % 5.

If I want to divide 6m into 5 equally long pieces, I would just do 6m / 5 = 1.2m ? There is no leftover?
I think what you are suggesting implies some assumption that you want to make one piece of 5m, and then divide that into 5 pieces of 1m? But then if you want pieces of 1m I would just do 6m % 1m ?

Also, how do you interpret something like 6 % 2.5 (when the rhs is not an integer)? I understand it as: I want to make pieces of 2.5 in size, how much am I left with at the end?

**Comment by jl-wynen** [comment]
> If I want to divide 6m into 5 equally long pieces, I would just do 6m / 5 = 1.2m ? There is no leftover?

True. What I meant was equally long pieces assuming that you can only cut at integer multiples of 1m. The reason is that mod is defined in terms of rounding to integers (in our case using `floor`). So the scale of the unit determines the smallest length we can divide into.

And this extends to non-integral rhs. The unintuitive bit here is that you can't distribute something to, e.g., 2.5 groups.

> In fact I have still no idea what the unit of that should be.

Same as the dividend according to the definition: `a == floor(a / b) * b + a % b`. Anything else would require fiddling with the maths. I'd be extremely careful with doing that just to satisfy some intuition.

**Comment by nvaytet** [comment]
> Same as the dividend according to the definition: a == floor(a / b) * b + a % b

I think that definition was introduced without physical units in mind.
I know it doesn't mean we should discard it, because it is consistent with the other math operations we have, but I don't think what we are talking about is just an intuition. To me, some of it just doesn't make much sense.

To try and bring other aspects/opinions into the discussion, I had a look at `pint`.
I know they may not be the reference, or correct (I still don't know what the correct answer is), but they forbid doing `6m % 5`, saying you cannot convert from dimensionless to meter.

**Comment by SimonHeybrock** [comment]
@jl-wynen Is there a reason why you *really* need to support mismatching units? We are not saying that it cannot be defined somehow, but it seems evident that it is confusing and not obvious.

**Comment by jl-wynen** [comment]
No, I don't have a concrete need. I just don't like limiting the operation beyond what is mathematically required. But if you think that this prevents mistakes, go ahead!

But note that I did not mean to argue for mismatching units in general but for supporting dimensionless and counts divisors regardless of the unit of the dividend.
